pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/FunctionConfigUtils.java

```diff
@@ -317,9 +319,15 @@ public static FunctionConfig convertFromDetails(FunctionDetails functionDetails)
         }
         if (functionDetails.getSource().getSubscriptionType() == Function.SubscriptionType.FAILOVER) {
             functionConfig.setRetainOrdering(true);
+            functionConfig.setRetainKeyOrdering(false);
             functionConfig.setProcessingGuarantees(FunctionConfig.ProcessingGuarantees.EFFECTIVELY_ONCE);
+        } else if (functionDetails.getSource().getSubscriptionType() == Function.SubscriptionType.KEY_SHARED) {
+            functionConfig.setRetainOrdering(false);
+            functionConfig.setRetainKeyOrdering(true);
+            functionConfig.setProcessingGuarantees(FunctionConfig.ProcessingGuarantees.ATLEAST_ONCE);
```

Comment:
We are we assuming ATLEAST_ONCE here?  A user can set the processing guarantee to be at most once and still set the retain ordering to be true

Reply:
Yes. Even the existing code is broken in that regard. I will fix that